src/basic/terminal-util.c

```diff
@@ -228,49 +231,261 @@ int ask_char(char *ret, const char *replies, const char *fmt, ...) {
+                goto fallback;
+
+        struct termios new_termios = old_termios;
+        termios_disable_echo(&new_termios);
+        if (tcsetattr(fd_input, TCSADRAIN, &new_termios) < 0)
+                return -errno;
+
+        for (;;) {
+                int c = fgetc(stdin);
+
+                /* On EOF or NUL, end the request, don't output anything anymore */
+                if (IN_SET(c, EOF, 0))
+                        break;
+
+                /* On Return also end the request, but make this visible */
+                if (IN_SET(c, '\n', '\r')) {
+                        fputc('\n', stdout);
+                        break;
+                }
+
+                if (c == '\t') {
+                        /* Tab */
+
+                        _cleanup_strv_free_ char **completions = NULL;
+                        if (get_completions) {
+                                r = get_completions(string, &completions, userdata);
+                                if (r < 0)
+                                        goto fail;
+                        }
+
+                        _cleanup_free_ char *new_string = NULL;
+                        CompletionResult cr = pick_completion(string, completions, &new_string);
+                        if (cr < 0) {
+                                r = cr;
+                                goto fail;
+                        }
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_FULL)) {
+                                /* Output the new suffix we learned */
+                                fputs(ASSERT_PTR(startswith(new_string, strempty(string))), stdout);
+
+                                /* And update the whole string */
+                                free_and_replace(string, new_string);
+                                n = strlen(string);
+                        }
+                        if (cr == COMPLETION_NONE)
+                                fputc('\a', stdout); /* BEL */
+
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_ALREADY)) {
+                                /* If this worked only partially, or if the user hit TAB even though we were
+                                 * complete already, then show the full list. */
+                                fputc('\n', stdout);
+
+                                STRV_FOREACH(nc, completions) {
+                                        const char *e;
+
+                                        e = startswith(*nc, strempty(string));
+                                        if (!e)
+                                                continue;
+
+                                        printf("\t%s%.*s%s%s\n",
+                                               ansi_grey(),
+                                               (int) (e - *nc), *nc,
+                                               ansi_normal(),
+                                               e);
+                                }
```

Comment:
So this does not make good use of the limited screen space available in the linux console in which we cannot scroll. We should use `show_menu()` or something similar here to show the completions so that we can fit more of them on the screen

Reply:
Or even better, make showing the completions a callback as well so that we can do it from the caller side.

Reply:
the menu in the caller side is actually a bit different thatn the one we want to show here. the caller site wants to shown the numbers, so that you can type a number instead of a name if you like. for completion this makes little sense, as the numbers cannot really be typed in here. otoh for the completion it makes sense to indicate what is already typed and what the completion options are in the output (i.e. what happens here with the ansi_grey() stuff), which doesn't make sense for the other menu.

hence, I think the menus should be similar, but also different.